File: src/typescript/3_push_pull/sink.ts
import * as zmq from "zeromq";
import { performance } from 'perf_hooks';

async function runSink() {
  console.log("Initializing sink...");
  const context = new zmq.Context();

  // Socket to receive messages on
  const receiver = new zmq.Pull();
  await receiver.bind("tcp://*:5558");
  console.log("Sink bound to tcp://*:5558");

  console.log("Sink ready!");

  // Wait for start of batch signal from ventilator
  const [startSignal] = await receiver.receive();
  console.log("Received start signal from ventilator:", startSignal.toString());

  // Start our clock now
  console.log("Starting clock...");
  const tstart = performance.now();

  // Process 100 confirmations
  const numTasks = 100;
  for (let task_nbr = 0; task_nbr < numTasks; task_nbr++) {
    const [confirmation] = await receiver.receive(); // Receive confirmation
    if (task_nbr % 10 === 0) {
      process.stdout.write(':');
      // console.log(`\nReceived confirmation number ${task_nbr + 1}`); // Optional detail
    } else {
      process.stdout.write('.');
    }
  }
  process.stdout.write('\n'); // Newline after dots

  // Calculate and report duration of batch
  const tend = performance.now();
  const duration = tend - tstart;
  console.log(`Total elapsed time: ${duration.toFixed(2)} msec`);

  console.log("Sink finished.");
  receiver.close();
  // context.close(); // Close context if needed
}

runSink().catch(err => {
  console.error("Sink error:", err);
  process.exit(1);
}); 

File: src/python/chapter_1/3_push_pull/sink.py
# Task sink
# Binds PULL socket to tcp://localhost:5558
# Collects results from workers via that socket
#
# Author: Lev Givon <lev(at)columbia(dot)edu>

import sys
import time
import zmq

print("Initializing sink...")
context = zmq.Context()

# Socket to receive messages on
receiver = context.socket(zmq.PULL)
receiver.bind("tcp://*:5558")

print("Sink ready!")

# Wait for start of batch
s = receiver.recv()

# Start our clock now
print("Starting clock...")
tstart = time.time()

# Process 100 confirmations
for task_nbr in range(100):
    s = receiver.recv()
    if task_nbr % 10 == 0:
        sys.stdout.write(':')
        print(f"Received confirmation number {task_nbr}")
    else:
        sys.stdout.write('.')
    sys.stdout.flush()

# Calculate and report duration of batch
tend = time.time()
print(f"Total elapsed time: {(tend-tstart)*1000} msec")

print("Done!")